dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/internal/VariantIterator.kt

```diff
@@ -0,0 +1,35 @@
+package com.datadog.gradle.plugin.internal
+
+/**
+ * This Iterator will take a list of flavor name and iterate over all possible
+ * (partial) variants from those.
+ *
+ * The partial variants will be iterated over in increasing order of priority.
+ * Product flavors that belong to the first flavor dimension have a higher priority than
+ * those belonging to the second flavor dimension.
+ *
+ * E.g.: for the variant with flavor names ["pro", "green", "europe"], the iterator will
+ * return the following values (in order):
+ * "europe", "green", "greenEurope", "pro", "proEurope", "proGreen", "proGreenEurope"
+ */
+internal class VariantIterator(
+    private val names: List<String>
+) : Iterator<String> {
+
+    private var index = 1
+    private val max = 1 shl names.size
+
+    override fun hasNext(): Boolean {
+        return index < max
+    }
+
+    override fun next(): String {
+        if (index >= max) {
+            throw NoSuchElementException()
+        }
+        val mask = index.toString(2).padStart(names.size, '0').map { it == '1' }
```

Comment:
nit: I would suggest to add explanation/example of evaluation as a comment

Reply:
Or maybe a simplified version:
```
internal class VariantIterator(
    private val names: List<String>
) : Iterator<String> {

    private var index = 0
    private var variants: MutableList<String> = mutableListOf()

    init {
        addVariant("")
    }

    private fun addVariant(prefix: String = "", level: Int = 0) {
        for (i in level until names.size) {
            val variant = prefix + if (level != 0) {
                names[i].capitalize()
            } else {
                names[i]
            }
            variants.add(variant)
            addVariant(variant, i + 1)
        }
    }

    override fun hasNext(): Boolean {
        return index < variants.size
    }

    override fun next(): String {
        if (index >= variants.size) {
            throw NoSuchElementException()
        }
        return variants[index++]
    }
}
```

Reply:
Nice trick Marius, but the idea of this implementation was to avoid generating and storing the full list of all possible combinations (which would be exponential). Also the implementation you gave doesn't provide the correct order of all combinations.